### PR TITLE
Support `@[Deprecated]` on constants

### PR DESCRIPTION
**Pull request by HertzDevil**
Resolves part of #11043. Deprecated types would be done as part of #10125 because alias deprecations must be treated differently.

```crystal
@[Deprecated]
FOO = 1

{{ FOO }}      # Warning: Deprecated FOO.
FOO            # Warning: Deprecated FOO.
x : Int32[FOO] # Warning: Deprecated FOO.

class C(N); end
module M(N); end

class Foo < C(FOO) # Warning: Deprecated FOO.
  include M(FOO)   # Warning: Deprecated FOO.
end
```

Because constants are internally represented as types, it is already possible to attach annotations to them, but constants are never exposed as `TypeNode`s in the macro language (e.g. `Path#resolve` and `TypeNode#constant` return the constant's initializer directly), so annotations on constants other than `@[Deprecated]` are still ignored.

### CODE REVIEW

File: src/compiler/crystal/semantic/top_level_visitor.cr [hunk 1]
```diff
@@ -793,6 +793,10 @@ class Crystal::TopLevelVisitor < Crystal::SemanticVisitor
     const = Const.new(@program, scope, name, value)
     const.private = true if target.visibility.private?
 
+    process_annotations(annotations) do |annotation_type, ann|
+      const.add_annotation(annotation_type, ann) if annotation_type == @program.deprecated_annotation
```
Comment (straight-shoota) [inline src/compiler/crystal/semantic/top_level_visitor.cr:797]: Please add a code comment why only `Deprecated` is supported here (maybe just a link to this PR?).

**Review by straight-shoota — COMMENTED** [review]
I'd prefer to do the refactoring *first* and in a separate PR. It clearly separates refactoring from feature addition. And it's easier to review and traceable. Could you extract that please?

File: src/compiler/crystal/semantic/warnings.cr [hunk 2]
```diff
@@ -48,6 +49,30 @@ module Crystal
...
+         (deprecated_annotation = DeprecatedAnnotation.from(ann))
+        call_location = node.location.try(&.macro_location) || node.location
+
+        return if self.ignore_warning_due_to_location?(call_location)
+        warning_key = call_location.try { |l| "#{const} #{l}" }
+
+        # skip warning if the call site was already informed
+        # if there is no location information just inform it.
+        return if !warning_key || @deprecated_constants_detected.includes?(warning_key)
+        @deprecated_constants_detected.add(warning_key) if warning_key
```
Comment (straight-shoota) [inline src/compiler/crystal/semantic/warnings.cr:65]: This early return condition appears to contradict the comment? The logic doesn't seem to make sense. But this is apparently copied from existing deprecation handlers, to I presume it's better to fix this outside this PR. I don't mind whether before or after.
In any case, I presume there could also be some more refactoring such as extracing common functionality from these methods.

**Review by straight-shoota — APPROVED** [review]

File: src/compiler/crystal/semantic/warnings.cr (hunk 2 above)
Comment (straight-shoota) [reply src/compiler/crystal/semantic/warnings.cr:65]: @HertzDevil Did you see this comment?

**Review by straight-shoota — APPROVED** [review]

**Review by bcardiff — APPROVED** [review]

**Pull request merged**